contracts/src/token/erc20/extensions/erc4626.rs

```diff
@@ -0,0 +1,403 @@
+        if shares > max_shares {
+               return Err(Error::ExceededMaxMint(ERC4626ExceededMaxMint{
+                receiver,
+                shares,
+                max: max_shares
+            }));
+        }
+        let assets  = self.preview_mint(shares);
+        self._deposit(msg::sender(), receiver, assets, shares)?;
+        Ok(assets)
+    }
+    
+/// Returns the maximum amount of assets that can be withdrawn by the given owner.
+/// 
+/// # Parameters
+/// - `owner`: The address of the account requesting to withdraw assets.
+///
+/// # Returns
+/// The maximum number of assets that can be withdrawn, which is `U256::MAX`.
+
+    fn max_withdraw(&self,owner: Address) -> U256 {
+        todo!()
+    }
+    
+    fn preview_withdraw( &self,assets: U256) -> U256 {
+        todo!()
+    }
+    
+    fn withdraw(&mut self,assets: U256, receiver: Address, owner: Address) -> Result<U256, Error> {
+        let  max_assets = self.max_withdraw(owner);
+        if assets > max_assets {
+               return Err(Error::ExceededMaxWithdraw(ERC4626ExceededMaxWithdraw{
+                owner,
+                assets,
+                max: max_assets
+            }));
+        }
+
+        let  shares = self.preview_redeem(assets);
+        self._withdraw(msg::sender(), receiver, owner, assets, shares)?;
+        Ok(assets)
+    }
+    
+    fn max_redeem(&self,owner: Address) -> U256 {
+        self._asset.balance_of(owner)
+    }
+    
+    fn preview_redeem(&self,shares: U256) -> U256 {
+        todo!()
+    }
+    
+    fn redeem(&mut self,shares: U256, receiver: Address, owner: Address) -> Result<U256, Error> {
+        let  max_shares = self.max_redeem(owner);
+        if shares > max_shares {
+               return Err(Error::ExceededMaxRedeem(ERC4626ExceededMaxRedeem{ 
+                owner, shares,
+                max: max_shares
+            }));
+        }
+
+        let  assets = self.preview_redeem(shares);
+        self._withdraw(msg::sender(), receiver, owner, assets, shares)?;
+        Ok(assets)
+    }
+}
+
+
+impl  ERC4626 {
+       fn _convert_to_shares(&self, assets: U256, rounding:bool) -> U256  {
+         let confersion_factor = self._asset._total_supply.get()  + U256::from(self._decimals_offset());
+         let dominitor = self.total_assets() + U256::from(1);
+         assets * confersion_factor / dominitor
+       }
+
+       fn _convert_to_assets(&self, shares: U256, rounding:bool) -> U256  { 
+         let confersion_factor = self._asset._total_supply.get()  + U256::from(self._decimals_offset());
+         let dominitor = self.total_assets() + U256::from(1);
+         shares * confersion_factor / dominitor
+       }
+
+        fn _deposit(&mut self,caller:Address,receiver:Address, assets:U256, shares:U256) -> Result<(), Error> {
+            // _SafeERC20.safeTransferFrom(_asset, caller, address(this), assets);
+            self._asset._mint(receiver, shares)?;
```

Comment:
Be careful, you are not calling `_mint` on the _asset_, but on the _contract itself_ (see [Solidity version](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC4626.sol#L248)).

Maybe it would be helpful to take inspiration from the [Erc20FlashMint extension PR](https://github.com/OpenZeppelin/rust-contracts-stylus/pull/407/files#diff-6548599a00b45c200a28b1fb023de5279f94a29c6ee35d7b614f4e7acedbead1R294), as it has a similar design problem as ERC4626, namely that you need to extend ERC20, but still leverage ERC20's functions (which is a problem in Stylus due to lack of proper inheritance).

One way to solve this is to make each ERC4626 function that needs to call base contract's functions (e.g. `_mint`) accept an `erc20: &mut Erc20` parameter, which acts as a reference to the base/parent contract.

Note that you'll still need `_asset: Erc20`.
But you can make it a generic `IErc20` type (as is done in [Solidity](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC20/extensions/ERC4626.sol#L51)) (@bidzyyys @qalisander wdyt about this approach?):

```rust
#[storage] // this is the new convention we use for defining storage
#[allow(clippy::pub_underscore_fields)]
pub struct ERC4626<T: IErc20 + StorageType> {
    pub _asset: T, 
    pub _underlying_decimals: StorageU8,
}
```

NOTE: although `_asset` and `_underlying_decimals` should be private, we're making public for the purposes of testing. This will be changed at some point.

Reply:
Thanks you

Reply:
Ignore this comment, as the actual needs of the contract require a different design.
See https://github.com/OpenZeppelin/rust-contracts-stylus/pull/465#discussion_r1898444318